proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+# Move Function + "Use After Move" Diagnostic
+
+* Proposal: [SE-NNNN](nnnn-move-function.md)
+* Authors: [Michael Gottesman](https://github.com/gottesmm), [Andrew Trick](https://github.com/atrick)
+* Review Manager: TBD
+* Status: Implemented on main as stdlib SPI (`_move` instead of `move`)
+
+<!--
+*During the review process, add the following fields as needed:*
+
+* Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN) or [apple/swift-evolution-staging#NNNNN](https://github.com/apple/swift-evolution-staging/pull/NNNNN)
+* Decision Notes: [Rationale](https://forums.swift.org/), [Additional Commentary](https://forums.swift.org/)
+* Bugs: [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN), [SR-MMMM](https://bugs.swift.org/browse/SR-MMMM)
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+-->
+
+## Introduction
+
+In this document, we propose adding a new function called `move` to the swift
+standard library, which ends the lifetime of a specific local `let`,
+local `var`, or `consuming` parameter. In order to enforce this, the compiler will
+emit a flow-sensitive diagnostic upon any uses that are after the move
+function. As an example:
+
+```
+useX(x) // do some stuff with local variable x
+
+// Ends lifetime of x, y's lifetime begins.
+let y = move(x) // [1]
+
+useY(y) // do some stuff with local variable y
+useX(x) // error, x's lifetime was ended at [1]
+
+// Ends lifetime of y. Since _ is no-op, we perform an actual release here.
```

Comment:
Whether there is a "release" depends on the type of `x`.  Focus on the language implications:
```
  Ends the lifetime of the variable `y`.  Assigning to `_` avoids creation of a new lifetime.
```
I would also through here be careful to distinguish the lifetime of a "variable" from the lifetime of the "value" that may be temporarily stored in that value.